File: packages/server/src/db/schema/Quiz.ts
import { Entity, EntityProperty, Index, ManyToOne, PrimaryKey, Property, Type, Unique } from '@mikro-orm/core'
import dayjs from 'dayjs'
import { Ulid } from 'id128'

import { getNextReview, repeatReview, srsMap } from '../quiz'
import { Note } from './Note'
import { Template } from './Template'
import { User } from './User'

export class DateTimeType extends Type<Date, string> {
  convertToDatabaseValue (value: Date | string): string {
    return dayjs(value).toISOString()
  }

  convertToJSValue (value: Date | string): Date {
    return dayjs(value).toDate()
  }

  getColumnType (prop: EntityProperty): string {
    return prop.length ? `DATE(${prop.length})` : 'DATE'
  }
}

export interface IQuiz {
  user: User
  note: Note
  template: Template
  front?: string
  back?: string
  mnemonic?: string
}

@Entity()
@Unique({ properties: ['user', 'note', 'template'] })
export class Quiz implements IQuiz {
  /*
   * Base Properties
   */

  @PrimaryKey()
  _id: string = Ulid.generate().toCanonical()

  @ManyToOne({ nullable: false })
  user!: User

  @ManyToOne({ nullable: false })
  note!: Note

  @ManyToOne({ nullable: false })
  template!: Template

  @Property({ lazy: true })
  front?: string

  @Property({ lazy: true })
  back?: string

  @Property({ lazy: true })
  mnemonic?: string

  /*
   * Quiz Properties
   */

  @Property()
  @Index()
  srsLevel?: number

  @Property({ type: DateTimeType })
  @Index()
  nextReview?: Date

  @Property()
  rightStreak?: number

  @Property()
  @Index()
  wrongStreak?: number

  @Property()
  maxRight?: number

  @Property()
  @Index()
  maxWrong?: number

  @Property({ type: DateTimeType })
  lastRight?: Date

  @Property({ type: DateTimeType })
  @Index()
  lastWrong?: Date

  constructor (d: IQuiz) {
    Object.assign(this, d)
  }

  markRight (): void {
    return this._updateSrsLevel(+1)()
  }

  markWrong (): void {
    return this._updateSrsLevel(-1)()
  }

  markRepeat (): void {
    return this._updateSrsLevel(0)()
  }

  private _updateSrsLevel (dSrsLevel: number) {
    return () => {
      this.rightStreak = this.rightStreak || 0
      this.wrongStreak = this.wrongStreak || 0
      this.maxRight = this.maxRight || 0
      this.maxWrong = this.maxWrong || 0

      if (dSrsLevel > 0) {
        this.rightStreak = this.rightStreak + 1
        this.wrongStreak = 0
        this.lastRight = new Date()

        if (this.rightStreak > this.maxRight) {
          this.maxRight = this.rightStreak
        }
      } else if (dSrsLevel < 0) {
        this.wrongStreak = this.wrongStreak + 1
        this.rightStreak = 0
        this.lastWrong = new Date()

        if (this.wrongStreak > this.maxWrong) {
          this.maxWrong = this.wrongStreak
        }
      }

      this.srsLevel = this.srsLevel || 0

      this.srsLevel += dSrsLevel

      if (this.srsLevel >= srsMap.length) {
        this.srsLevel = srsMap.length - 1
      }

      if (this.srsLevel < 0) {
        this.srsLevel = 0
      }

      if (dSrsLevel > 0) {
        this.nextReview = getNextReview(this.srsLevel)
      } else {
        this.nextReview = repeatReview()
      }
    }
  }
}

File: packages/j-server/src/main/kotlin/rep2recall/db/Quiz.kt
package rep2recall.db

import com.github.guepardoapps.kulid.ULID
import org.jetbrains.exposed.dao.Entity
import org.jetbrains.exposed.dao.EntityClass
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.sql.jodatime.datetime
import org.joda.time.DateTime
import org.joda.time.Duration

object QuizTable: IdInitTable<String>("quiz") {
    override val id = varchar("id", 26).entityId()
    val updatedAt = datetime("updated_at").nullable()
    val userId = reference("user_id", UserTable)

    val noteId = reference("note_id", NoteTable).nullable()
    val templateId = reference("template_id", TemplateTable).nullable()

    val deck = varchar("deck", 200)
    val front = varchar("front", 1000).nullable()
    val back = varchar("back", 1000).nullable()
    val mnemonic = varchar("mnemonic", 1000).nullable()

    val srsLevel = integer("srs_level").nullable().index()
    val nextReview = datetime("next_review").nullable().index()
    val rightStreak = integer("right_streak").nullable().index()
    val wrongStreak = integer("wrong_streak").nullable().index()
    val maxRight = integer("max_right").nullable().index()
    val maxWrong = integer("max_wrong").nullable().index()
    val lastRight = datetime("last_right").nullable().index()
    val lastWrong = datetime("last_wrong").nullable().index()

    override fun init() {
        uniqueIndex(userId, noteId, templateId)
    }
}

class Quiz(id: EntityID<String>): Entity<String>(id) {
    companion object: EntityClass<String, Quiz>(QuizTable) {
        override fun new(id: String?, init: Quiz.() -> Unit) = super.new(id ?: ULID.random(), init)
        override fun new(init: Quiz.() -> Unit) = new(null, init)

        fun create(
                user: User,
                note: Note?,
                template: Template?,
                deck: String,
                front: String? = null,
                back: String? = null,
                mnemonic: String? = null,
                srsLevel: Int? = null,
                nextReview: DateTime? = null,
                rightStreak: Int? = null,
                wrongStreak: Int? = null,
                maxRight: Int? = null,
                maxWrong: Int? = null,
                lastRight: DateTime? = null,
                lastWrong: DateTime? = null,
                id: String? = null
        ): Quiz {
            return  new(id) {
                userId = user.id
                noteId = note?.id
                templateId = template?.id
                this.deck = deck
                this.front = front
                this.back = back
                this.mnemonic = mnemonic
                this.srsLevel = srsLevel
                this.nextReview = nextReview
                this.rightStreak = rightStreak
                this.wrongStreak = wrongStreak
                this.maxRight = maxRight
                this.maxWrong = maxWrong
                this.lastRight = lastRight
                this.lastWrong = lastWrong
            }
        }

        val srsMap = listOf(
                Duration.standardHours(1),
                Duration.standardHours(4),
                Duration.standardHours(8),
                Duration.standardDays(1),
                Duration.standardDays(3),
                Duration.standardDays(1 * 7),
                Duration.standardDays(2 * 7),
                Duration.standardDays(4 * 7),
                Duration.standardDays(16 * 7)
        )
    }

    var updatedAt by QuizTable.updatedAt
    var userId by QuizTable.userId
    val user by User referencedOn QuizTable.userId

    var noteId by QuizTable.noteId

    var templateId by QuizTable.templateId
    val template by Template optionalReferencedOn QuizTable.templateId

    var deck: String by QuizTable.deck
    var front by QuizTable.front
    var back by QuizTable.back
    var mnemonic by QuizTable.mnemonic

    var srsLevel by QuizTable.srsLevel
    var nextReview by QuizTable.nextReview
    var rightStreak by QuizTable.rightStreak
    var wrongStreak by QuizTable.wrongStreak
    var maxRight by QuizTable.maxRight
    var maxWrong by QuizTable.maxWrong
    var lastRight by QuizTable.lastRight
    var lastWrong by QuizTable.lastWrong

    @Suppress("unused")
    fun markRight() = updateSrsLevel(1)
    @Suppress("unused")
    fun markWrong() = updateSrsLevel(-1)
    @Suppress("unused")
    fun markRepeat() = updateSrsLevel(0)

    private fun updateSrsLevel(dSrsLevel: Int) {
        updatedAt = DateTime.now()

        rightStreak = rightStreak ?: 0
        wrongStreak = wrongStreak ?: 0
        maxRight = maxRight ?: 0
        maxWrong = maxWrong ?: 0

        if (dSrsLevel > 0) {
            rightStreak = rightStreak!! + 1
            wrongStreak = 0
            lastRight = DateTime.now()

            if (rightStreak!! > maxRight!!) {
                maxRight = rightStreak
            }
        } else if (dSrsLevel < 0) {
            wrongStreak = wrongStreak!! + 1
            rightStreak = 0
            lastWrong = DateTime.now()

            if (wrongStreak!! > maxWrong!!) {
                maxWrong = wrongStreak
            }
        }

        srsLevel = (srsLevel ?: 0) + dSrsLevel

        if (srsLevel!! >= srsMap.size) {
            srsLevel = srsMap.size - 1
        }

        if (srsLevel!! < 0) {
            srsLevel = 0
        }

        nextReview = if (dSrsLevel > 0) {
            DateTime.now()
                    .plus(srsMap.elementAtOrElse(srsLevel!!) { Duration.standardHours(4) })
        } else {
            DateTime.now()
                    .plus(Duration.standardHours(1))
        }
    }
}